NetKet/Graph/lattice_impl.hpp

```diff
@@ -0,0 +1,430 @@
+              is_bipartite = false;
+            }
+          }
+        });
+  }
+  return is_bipartite;
+}
+
+// Lattice representations (site = k, vector = n_i, coord = cartesian
+// coordinates)
+std::vector<int> Lattice::Site2Vector(int i) const {
+  int ndim = extent_.size();
+  std::vector<int> result(ndim, 0);
+  int ip;
+  if (i < nlatticesites_)
+    ip = i;
+  else
+    ip = i % nlatticesites_;
+  int k = ndim - 1;
+  while (ip > 0) {
+    result[k] = ip % extent_[k];
+    ip /= extent_[k];
+    k--;
+  }
+  return result;
+};
+std::vector<double> Lattice::Vector2Coord(const std::vector<int> &n,
+                                          int iatom) const {
+  std::vector<double> R(ndim_, 0);
+  for (int j = 0; j < ndim_; j++) {
+    R[j] = atoms_coord_[iatom][j];
+    for (int k = 0; k < ndim_; k++) R[j] += n[k] * basis_vectors_[k][j];
+  }
+
+  return R;
+};
+std::vector<double> Lattice::Site2Coord(int k) const { return R_[k]; }
+int Lattice::Vector2Site(const std::vector<int> &n) const {
+  int k = 0;
+
+  for (int i = 0; i < Ndim(); i++) {
+    int base = 1;
+    for (int j = i + 1; j < Ndim(); j++) base *= extent_[j];
+    k += n[i] * base;
+  }
+
+  return k;
+}
+
+// Neighbours
+std::vector<std::vector<int>> Lattice::PossibleLatticeNeighbours() const {
+  const int nsymbols = 3;
+  const int max = nsymbols - 1;
+  std::vector<int> b(ndim_, 0);
+  std::vector<int> result_vector;
+  std::vector<std::vector<int>> result_matrix;
+  std::vector<int> row(ndim_, 0);
+  do {
+    for (std::vector<int>::const_iterator ib = b.begin(); ib != b.end(); ++ib)
+      result_vector.push_back(*ib);
+  } while (next_variation(b.begin(), b.end(), max));
+
+  for (std::size_t i = 0; i < result_vector.size() / ndim_; i++) {
+    for (int j = 0; j < ndim_; j++) {
+      row[j] = result_vector[i * ndim_ + j] -
+               1;  //-1 because I want (-1,0,1) as symbols instead of (0,1,2)
+    }
+    result_matrix.push_back(row);
+  }
+  return result_matrix;
+};
+std::vector<double> Lattice::NeighboursSquaredDistance(
+    const std::vector<std::vector<int>> &neighbours_matrix, int iatom) const {
+  std::vector<double> distance;
+  for (std::size_t row = 0; row < neighbours_matrix.size(); row++) {
+    std::vector<int> n(ndim_, 0);
+    for (int j = 0; j < ndim_; j++) n[j] = neighbours_matrix[row][j];
+    for (int jatom = 0; jatom < natoms_; jatom++)
+      distance.push_back(
+          GetSquaredDistance(Vector2Coord(n, jatom), atoms_coord_[iatom]));
+  }
+
+  return distance;
+};
+std::vector<std::vector<int>> Lattice::LatticeNeighbours(int iatom) const {
+  std::vector<std::vector<int>> neighbours_matrix_in;
+  std::vector<std::vector<int>> result;
+  std::vector<double> distance;
+  double min_distance;
+
+  neighbours_matrix_in = PossibleLatticeNeighbours();
+
+  distance = NeighboursSquaredDistance(neighbours_matrix_in, iatom);
+
+  min_distance = *min_nonzero_elem(distance.begin(), distance.end());

```

Comment:
This looks like undefined behaviour. Your implementation of `min_nonzero_elem` can return `distance.end()`. If you're absolutely sure it can't happen, consider adding an assert either here or in `min_nonzero_elem`.